### PR TITLE
Refactor variables

### LINKED ISSUE #235: `gather` is broken for purposes of training
Consider the MWE:
```julia

@testset "Check gather can be used in a actual network" begin
    let
        sess3 = Session(Graph())
        x = get_variable("x", (50,10), Float64)
        x3n5 = gather(x, [3, 5])
        cost = reduce_sum(x3n5)
        optimizer = train.minimize(train.AdamOptimizer(0.1), cost)
        run(sess3, global_variables_initializer())
        @test size(run(sess3, x3n5)) == (2, 10)
    end
end

```
Gives me the error:
```
Check gather can be used in a actual network: Error During Test
  Got an exception of type MethodError outside of a @test
  MethodError: Cannot `convert` an object of type TensorFlow.Operation to an object of type TensorFlow.RawTensor
  This may have arisen from a call to the constructor TensorFlow.RawTensor(...),
  since type constructors fall back to convert methods.
   in (::TensorFlow.Ops.##352#353{TensorFlow.Operation,DataType})() at /home/ubuntu/.julia/v0.5/TensorFlow/src/ops/imported_ops.jl:6085
   in with_op_name(::TensorFlow.Ops.##352#353{TensorFlow.Operation,DataType}, ::Void, ::String) at /home/ubuntu/.julia/v0.5/TensorFlow/src/core.jl:869
   in #const_#351(::Void, ::TensorFlow.Operation, ::Type{T}, ::Function) at /home/ubuntu/.julia/v0.5/TensorFlow/src/ops/imported_ops.jl:6076
   in (::TensorFlow.Ops.#kw##const_)(::Array{Any,1}, ::TensorFlow.Ops.#const_) at ./<missing>:0
   in #constant#135(::Void, ::Array{Any,1}, ::Function, ::TensorFlow.Operation) at /home/ubuntu/.julia/v0.5/TensorFlow/src/ops/sequences.jl:14
   in convert(::Type{TensorFlow.Tensor{Any}}, ::TensorFlow.Operation) at /home/ubuntu/.julia/v0.5/TensorFlow/src/core.jl:1119
   in (::TensorFlow.Ops.##1437#1438{Void})() at /home/ubuntu/.julia/v0.5/TensorFlow/src/ops/imported_ops.jl:25253
   in with_op_name(::TensorFlow.Ops.##1437#1438{Void}, ::Void, ::String) at /home/ubuntu/.julia/v0.5/TensorFlow/src/core.jl:869
   in #scatter_sub#1436(::Void, ::Void, ::Function, ::TensorFlow.Operation, ::TensorFlow.Tensor{Int64}, ::TensorFlow.Tensor{Float64}) at /home/ubuntu/.julia/v0.5/TensorFlow/src/ops/imported_ops.jl:25249
   in #apply_gradients#15(::Void, ::String, ::Function, ::TensorFlow.train.AdamOptimizer, ::Array{Tuple{Any,Any},1}) at /home/ubuntu/.julia/v0.5/TensorFlow/src/train.jl:176
   in (::TensorFlow.train.#kw##apply_gradients)(::Array{Any,1}, ::TensorFlow.train.#apply_gradients, ::TensorFlow.train.AdamOptimizer, ::Array{Tuple{Any,Any},1}) at ./<missing>:0
   in #minimize#1(::Void, ::Void, ::Void, ::Function, ::TensorFlow.train.AdamOptimizer, ::TensorFlow.Tensor{Float64}) at /home/ubuntu/.julia/v0.5/TensorFlow/src/train.jl:41
   in macro expansion; at ./REPL[3]:7 [inlined]
   in macro expansion; at ./test.jl:674 [inlined]
   in anonymous at ./<missing>:?
   in eval(::Module, ::Any) at ./boot.jl:234
   in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:64
   in macro expansion at ./REPL.jl:95 [inlined]
   in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:68
Test Summary:                                | Error  Total
  Check gather can be used in a actual network |     1      1
ERROR: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
 in finish(::Base.Test.DefaultTestSet) at ./test.jl:498
 in macro expansion; at ./test.jl:681 [inlined]
 in anonymous at ./<missing>:?
```

As a workaround I added:

 function Base.convert{T}(::Type{Tensor{T}}, value::Operation)
    warn("convert{T}{::Tensor{T}, ::Operation} was called  on the operation $(value.name). This was ignored")
    value
end

And the output is:

```
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation x. This was ignored
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation x. This was ignored
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation adam/x/m. This was ignored
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation adam/x/m. This was ignored
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation adam/x/v. This was ignored
WARNING: convert{T}{::Tensor{T}, ::Operation} was called  on the operation adam/x/v. This was ignored
Test Summary:                                | Pass  Total
  Check gather can be used in a actual network |    1      1
```

I think the solution to this is for the code for generating ops, needs to special case inputs with the name `ref`.
Because right now, all input to operations get `v = convert(TensorFlow.Tensor{Any}, v)`

What I can't workout though, is why more things are not broken.

Eg I added

```julia
    result = run(sess, TensorFlow.scatter_sub(a, [1; 5], [1.; 1.]; use_locking=true))
    @test -ones(6) == result
```
to https://github.com/malmaud/TensorFlow.jl/blob/master/test/math.jl#L203
and all 6 assignments pass.

I think this has to do with `gradients` returning different types (`Operation`s) to what that is testing on (`Variable`s)

### PR DESCRIPTION
**Pull request by malmaud**
Closes #235

### CODE REVIEW

File: test/training.jl [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using Base.Test
```
Comment (oxinabox) [inline test/training.jl]: Maybe move [this testset](https://github.com/malmaud/TensorFlow.jl/blob/master/test/transformations.jl#L141-L165) here as well

File: src/variable.jl [hunk 2]
```diff
@@ -10,13 +29,19 @@ import .Ops:
 
 import Distributions
 
-type Variable <: AbstractTensor
-    var_node::Operation
-    assign_node::Operation
+type Variable{T} <: tf.AbstractTensor
```
Comment (oxinabox) [inline src/variable.jl]: I like adding {T} to `Variable`.
Since now all subtypes of `AbstractTensor` have it.
I think we should make it `AbstractTensor{T}`.
That will make a bunch of dispatch easier, eg in indexing.
And also will make it trivial to close #226

File: src/variable.jl [hunk 3]
```diff
@@ -1,4 +1,23 @@
-import .Ops:
+module Variables
```
Comment (oxinabox) [inline src/variable.jl]: Can you explain to me this increased use of modules?
They seem to make everything harder in julia.
(I wrote a [SO answer about just not using them, ages back](https://stackoverflow.com/a/40314833/179081))
I'm interested in the motivation here.

File: src/variable.jl [hunk 4]
```diff
@@ -39,40 +64,40 @@ running its `initializer` `Operation`, restoring the variable from a save file,
...
     return self
 end
 
-@with_def_graph function Variable(graph::Graph, s::AbstractString)
-    var = Variable()
-    var.var_node = get(get_node_by_name(graph, s))
-    var.assign_node = get(get_node_by_name(graph, "$s/Assign"))
-    var
+@tf.with_def_graph function Variable(graph::tf.Graph, s::AbstractString)
+    # var = Variable{Any}()
```
Comment (oxinabox) [inline src/variable.jl]: delete dead comment

File: src/variable.jl (hunk 3 above)
Comment (malmaud) [inline src/variable.jl]: It makes it a little easier to develop on since I can just re-evaluate the `Variables` module if I change any of the types in it. I also like not having too many global symbols in a single module.

File: src/variable.jl (hunk 2 above)
Comment (malmaud) [inline src/variable.jl]: Yes, we should have this. It can be its own separate PR though.

File: src/variable.jl (hunk 4 above)
Comment (malmaud) [inline src/variable.jl]: done

File: test/training.jl (hunk 1 above)
Comment (malmaud) [inline test/training.jl]: done
